### PR TITLE
Add another attempt to check write permissions on a folder (fixes #1971)

### LINKED ISSUE #1971: Can't set a share on the internal storage to "Send & Receive"
I can't set a share to "Send and Receive" through the app. It shows the following error:

WhatsApp Image 2023-09-06 at 15 27 09

The share is on my internal storage and not on the SD card. I can change this setting through the Web GUI though. Please let me know if you need any logs.

### Version Information

    Syncthing Version: v1.23.7
    Android Version: Android 13

### PR DESCRIPTION
**Pull request by wrobbins**
# Description
Fixes issue https://github.com/syncthing/syncthing-android/issues/1971

# Changes
* Add another attempt to check write permissions on a folder translating `~` to `externalStorageDirectory`. 
  * ignores `Constants.PREF_USE_ROOT` 
  * uses `java.io.File` to check access instead of shelling out.

## Notes
This an unobtrusive change, leaving the existing write-access check in place.

I tested this on a Pixel 6 Pro with Android 14.

### CODE REVIEW

File: app/src/main/java/com/nutomic/syncthingandroid/util/Util.java [hunk 1]
```diff
@@ -241,4 +272,5 @@ public static AlertDialog.Builder getAlertDialogBuilder(Context context)
     {
         return new MaterialAlertDialogBuilder(context);
     }
+
```
Comment (tomasz1986) [inline app/src/main/java/com/nutomic/syncthingandroid/util/Util.java:275]: Please remove the unrelated new line change.

File: app/src/main/java/com/nutomic/syncthingandroid/util/Util.java [hunk 2]
```diff
@@ -172,6 +171,38 @@ public static boolean nativeBinaryCanWriteToPath(Context context, String absolut
...
+    /**
+     * Returns if the syncthing binary would be able to write a file into
+     * the given folder given the configured access level.
+     *
+     * This uses the Android-native File API
+     */
+    public static Boolean nativeBinaryCanWriteToPath2(File externalStorageDir, String absoluteFolderPath) {
+        final String TOUCH_FILE_NAME = ".stwritetest";
+
+        // normalize path replacing ~ with external storage directory
```
Comment (tomasz1986) [inline app/src/main/java/com/nutomic/syncthingandroid/util/Util.java:183]:
```suggestion
        // Normalize path replacing ~ with external storage directory.
```

File: app/src/main/java/com/nutomic/syncthingandroid/util/Util.java [hunk 3]
```diff
@@ -172,6 +171,38 @@ public static boolean nativeBinaryCanWriteToPath(Context context, String absolut
...
+        // this is consistent with what $HOME is set to in SyncthingRunnable
```
Comment (tomasz1986) [inline app/src/main/java/com/nutomic/syncthingandroid/util/Util.java:184]:
```suggestion
        // This is consistent with what $HOME is set to in SyncthingRunnable.
```

File: app/src/main/java/com/nutomic/syncthingandroid/util/Util.java [hunk 4]
```diff
@@ -172,6 +171,38 @@ public static boolean nativeBinaryCanWriteToPath(Context context, String absolut
...
+        final String normalizedPath = absoluteFolderPath.replaceFirst("^~", externalStorageDir.getAbsolutePath());
+
+        // Write permission test file
```
Comment (tomasz1986) [inline app/src/main/java/com/nutomic/syncthingandroid/util/Util.java:187]:
```suggestion
        // Write permission test file.
```

File: app/src/main/java/com/nutomic/syncthingandroid/util/Util.java [hunk 5]
```diff
@@ -172,6 +171,38 @@ public static boolean nativeBinaryCanWriteToPath(Context context, String absolut
         return true;
     }
 
+    /**
+     * Returns if the syncthing binary would be able to write a file into
+     * the given folder given the configured access level.
+     *
+     * This uses the Android-native File API
```
Comment (tomasz1986) [inline app/src/main/java/com/nutomic/syncthingandroid/util/Util.java:178]:
```suggestion
     * This uses the Android-native File API.
```

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FolderActivity.java [hunk 6]
```diff
@@ -9,6 +9,8 @@
 import android.os.Build;
 import android.os.Bundle;
 import androidx.documentfile.provider.DocumentFile;
+
```
Comment (tomasz1986) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FolderActivity.java:12]: Please remove the new line.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FolderActivity.java [hunk 7]
```diff
@@ -537,6 +539,10 @@ private void checkWriteAndUpdateUI() {
          * Access level readwrite: folder can be configured "sendonly" or "sendreceive".
          */
         mCanWriteToPath = Util.nativeBinaryCanWriteToPath(FolderActivity.this, mFolder.path);
+        if(!mCanWriteToPath){
```
Comment (tomasz1986) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FolderActivity.java:542]:
```suggestion
        if (!mCanWriteToPath){
```

File: app/src/main/java/com/nutomic/syncthingandroid/util/Util.java [hunk 8]
```diff
@@ -172,6 +171,38 @@ public static boolean nativeBinaryCanWriteToPath(Context context, String absolut
...
+        final String TOUCH_FILE_NAME = ".stwritetest";
+
+        // normalize path replacing ~ with external storage directory
+        // this is consistent with what $HOME is set to in SyncthingRunnable
+        final String normalizedPath = absoluteFolderPath.replaceFirst("^~", externalStorageDir.getAbsolutePath());
+
+        // Write permission test file
+        File touchFile = new File(normalizedPath, TOUCH_FILE_NAME);
+        final String touchFilePath = touchFile.getAbsolutePath();
+        try{
```
Comment (tomasz1986) [inline app/src/main/java/com/nutomic/syncthingandroid/util/Util.java:190]:
```suggestion
        try {
```

Comment (tomasz1986) [reply app/src/main/java/com/nutomic/syncthingandroid/util/Util.java:190]: Please double-check whether the code follows the same convention as the current one (spacing, new lines, etc.).

Comment (wrobbins) [reply app/src/main/java/com/nutomic/syncthingandroid/util/Util.java]: @tomasz1986 thank you for the review. I addressed the formatting issues you highlighted.

File: app/src/main/java/com/nutomic/syncthingandroid/util/Util.java [hunk 9]
```diff
@@ -172,6 +171,38 @@ public static boolean nativeBinaryCanWriteToPath(Context context, String absolut
...
+     * the given folder given the configured access level.
+     *
+     * This uses the Android-native File API.
+     */
+    public static Boolean nativeBinaryCanWriteToPath2(File externalStorageDir, String absoluteFolderPath) {
+        final String TOUCH_FILE_NAME = ".stwritetest";
+
+        // Normalize path replacing ~ with external storage directory.
+        // This is consistent with what $HOME is set to in SyncthingRunnable.
+        final String normalizedPath = absoluteFolderPath.replaceFirst("^~", externalStorageDir.getAbsolutePath());
```
Comment (imsodin) [inline app/src/main/java/com/nutomic/syncthingandroid/util/Util.java:185]: Did you test explicitly if this part alone helped?

Seems like we should use the same environment, including setting `HOME`, when executing the shell command as we do in SyncthingRunnable.

**Review by imsodin — COMMENTED** [review]
Thanks for working on this! And sorry for the long wait on the review.  
This looks good. One question inline and a concern below, just in case you have any insight on it. 

I am a somewhat worried that Android handles IO differently when executing shell command vs using the java file API. However I think it's better to potentially have some false positives than the current state, where the detection always fails and users need to know they can workaround it by creating a folder in the web UI.

Also FYI it might be a while (at worst never) until this reaches users due to google, see https://github.com/syncthing/syncthing-android/issues/2064
